packages/grid/_modules_/grid/models/api/gridColumnApi.ts

```diff
@@ -47,8 +47,9 @@ export interface GridColumnApi {
   /**
    * Updates the definition of multiple columns at the same time.
    * @param {GridColDef[]} cols The new column [[GridColDef]] objects.
+   * @param {boolean} allColumnsProvided Determines if the `cols` argument contains all the columns or just the updated ones
    */
-  updateColumns: (cols: GridColDef[]) => void;
+  updateColumns: (cols: GridColDef[], allColumnsProvided?: boolean) => void;
```

Comment:
`allColumnsProvided` is this one to flag if it is a reset or an update? 🤔

Reply:
The issue was that @oliviertassinari was trying to use `updateColumns` when the column prop change.
But when we removed column from the prop, they were not removed from the state because `upsertColumnsState` keeps the previous columns in the internal state.
So I added a flag on `updateColumns` to ask the code not to keep the previous ones.

It's indeed a flag to differentiate update and reset

If you don't want to pollute `updateColumn`, we could add a method `setColumns` and the effect that runs when the `columns` prop changes would call `setColumns`. 
`updateColumns` could also call `setColumns` to keep as much logic in common as possible.

Reply:
> If you don't want to pollute updateColumn, we could add a method setColumns and the effect that runs when the columns prop changes would call setColumns.
updateColumns could also call setColumns to keep as much logic in common as possible.

Yes that sounds much cleaner.

Reply:
Updated